AP_HAL_ChibiOS: Add SaamPixV1_1 board support (#32582)
ArduPilot/ardupilot: libraries/AP_HAL_ChibiOS/hwdef/SaamPixV1_1/README.md

- PWM Outputs: 16 channels
- RC Input: PPM, SBUS, DSM/DSM2/DSMX via dedicated pin or UART
- USB: micro-USB (OTG1)
- Safety switch: not present

## Where to Buy

Available from [Saam Drones](https://www.saamdrones.com/).

## Pinout

![SaamPixV1_1 Pinout](SaamPixV1_1-pinout.png)

![SaamPixV1_1 Top View](SaamPixV1_1-top.png)

## UART Mapping

| Name | Pin | Function | Notes |
| ------ | ----- | ---------- | ------- |
| SERIAL0 | USB OTG1 | USB Console | |
| SERIAL1 | USART2 | Telemetry 1 | CTS/RTS on PD3/PD4 |
| SERIAL2 | USART3 | Telemetry 2 | |
| SERIAL3 | UART4 | GPS 1 | |
| SERIAL4 | UART8 | GPS 2 (default) | TX invertible via GPIO(78) |
| SERIAL5 | USART1 | SBUS input | |
| SERIAL6 | UART7 | Debug UART | |
| SERIAL7 | USART6 | RC Input alt | ALT config only |

Note: Some vendor silkscreens use compact labels such as "SBUS/RSSI" around the receiver area. In ArduPilot mapping these are separate signals: SBUS serial input is on USART1 RX (PB7), while RSSI input is the separate digital GPIO on PC1.

## CAN

The SaamPixV1_1 exposes one CAN interface:

- CAN1 RX: PD0
- CAN1 TX: PD1

## PWM Output

The SaamPixV1_1 has 16 PWM outputs. All outputs support PWM and DShot protocols.

> | Outputs | Timer | Pins | Notes |

Reviewer: cpu pins have no user value...remove this column

| --------- | ------- | ------ | ------- |
| 1-4 | TIM1 | PE9, PE11, PE13, PE14 | |
| 5-8 | TIM4 | PD12, PD13, PD14, PD15 | |
| 9-12 | TIM3 | PB4, PC7, PB0, PB1 | GPIO(50-53) |
| 13-16 | TIM2 | PA15, PB3, PA2, PA3 | GPIO(54-55) on 13-14 |

Note: All outputs within the same timer group must use the same protocol.

> Note: All 16 outputs are generated directly by FMU timers (no IOMCU on this board).

Reviewer: remove....unneeded

Note: Servo rail power (+V/GND on output headers) is supplied externally and is not generated or switched by the FMU.

## RC Input

> RC input is configured on PC6 by default using TIM8_CH1 and supports the following unidirectional protocols: PPM, SBUS, DSM, DSM2, DSMX, IBUS, SUM-D.

Reviewer: ```suggestion
RC input is configured on RC/PPM input by default and supports the following all ArduPilot compatible unidirectional protocols including PPM
```

> For SBUS, connect to SERIAL5 (USART1 RX on PB7). SBUS inverter control is on PA10.

Reviewer: not needed...nor the inverter control mention.....SBUS should be applied to RCIN/PPM

Reviewer: again, this is not needed...SBUS connects to RC/PPM input..

Author: Done — removed the inverter-control mention. The RC Input section now simply states that SBUS connects to the RC/PPM input.

> For bidirectional protocols (CRSF/ELRS, SRXL2, FPort), this board may require careful UART/DMA configuration and is not the recommended default RC path. Preferred default is TIM8 RCIN on PC6 or SBUS on USART1.

Reviewer: remove...CRSF,etc. are perfectly valid and recommended RC options

> An alternate RC input path via USART6 on PC6 is available using ALT config 1.

Reviewer: this wont work....since USART 6 has no DMA

> ### Receiver Connector Signal Mapping

Reviewer: remove this table...not needed...you have to get a UART with DMA and recommend it for bi-dir protocols...all uni-dir can use RC/PPM

| Connector Label | ArduPilot Signal | MCU Pin | Type | Notes |
| ----- | ----- | ----- | ----- | ----- |
| RCIN / PPM | TIM8 RCIN | PC6 | Digital input | Default RC input path |
| SBUS | USART1_RX (SERIAL5) | PB7 | UART RX | Requires receiver protocol setup |
| SBUS_INV_CTL | SBUS inverter control | PA10 | Digital output | Controls SBUS inversion only |
| RSSI | FMU_GPIO3 | PC1 | Digital input | Non-ADC RSSI input |
| UART8_TXINV | SERIAL4 TX inversion | PB12 (GPIO78) | Digital output | Affects UART8/SERIAL4 only |

This mapping is authoritative for the firmware and should be used in preference to compact silkscreen grouping labels.

> ## GPIOs

Reviewer: expand with GPIOs you need to add to other outputs

| GPIO | Pin | Function |
| ------ | ----- | ---------- |
| GPIO(0) | PA8 | RUN_LED |
| GPIO(32) | PA7 | ALARM (TIM14) |
| GPIO(50) | PB4 | Servo9 |
| GPIO(51) | PC7 | Servo10 |
| GPIO(52) | PB0 | Servo11 |
| GPIO(53) | PB1 | Servo12 |
| GPIO(54) | PA15 | Servo13 |
| GPIO(55) | PB3 | Servo14 |

> | GPIO(78) | PB12 | UART8 TX invert (SERIAL4 only, not SBUS) |

Reviewer: ```suggestion
| GPIO(78) | PB12 | UART8 TX signal output inversion control (not usually needed) |
```

> ## RSSI and Analog Pins

Reviewer: remove this section...no user value

[Saam Drones](https://www.saamdrones.com/). It features dual SPI IMUs,
an onboard barometer, dual compasses, CAN, microSD logging, and
RAMTRON parameter storage. It supports multirotor, fixed-wing, VTOL,
and hybrid UAV applications.

![SaamPixV1_1 Board](SaamPixV1_1-board.jpg)

## Features/Specifications

- MCU: STM32F427xx ARM Cortex-M4 with FPU at 168 MHz
- Flash: 2048 KB internal
- SRAM: 256 KB
- IMU 1: ICM20602
- IMU 2: MPU9250 (with onboard AK8963 compass)
- Barometer: MS5611
- Compass 1: LIS3MDL (external SPI)
- Compass 2: AK8963 (via MPU9250)
- RAMTRON parameter storage (16 KB)
- MicroSD card (SDIO interface)
- CAN bus: 1x CAN1
- UARTs: 6 (USART1, USART2, USART3, UART4, UART7, UART8)
  - USB OTG
- I2C: 2 buses (I2C1, I2C2)
- PWM Outputs: 16 channels
- RC Input: PPM, SBUS, DSM/DSM2/DSMX via dedicated pin or
  UART
- USB: micro-USB (OTG1)
- Safety switch: not present

## Where to Buy

Available from [Saam Drones](https://www.saamdrones.com/).

## Pinout

![SaamPixV1_1 Pinout](SaamPixV1_1-pinout.png)

![SaamPixV1_1 Top View](SaamPixV1_1-top.png)

## UART Mapping

| Name | Pin | Function | Notes |
| --- | --- | --- | --- |
| SERIAL0 | USB OTG1 | USB Console | |

> | SERIAL1 | USART2 | Telemetry 1 | CTS/RTS on PD3/PD4 |

Reviewer: ```suggestion
| SERIAL1 | USART2 | Telemetry 1/MAVLink2 | CTS/RTS on PD3/PD4 |
```

Author: Applied — the README now reads `Telemetry 1/MAVLink2`. Thanks.

> | SERIAL2 | USART3 | Telemetry 2 | |

Reviewer: ```suggestion
| SERIAL2 | USART3 | Telemetry 2/MAVLink2 | |
```

Author: On this board USART3 (PD8/PD9) is not an external telemetry header — it is wired to the onboard CP2104 USB-UART bridge on the secondary USB connector (verified against the schematic), intended for a companion computer. I've documented SERIAL2 as `Companion computer (USB)` so the README matches the actual wiring rather than implying an external Telem2 port. Happy to switch to the generic `Telemetry 2/MAVLink2` wording if you'd prefer it kept uniform.

| SERIAL3 | UART4 | GPS 1 | |
| SERIAL4 | UART8 | GPS 2 (default) | TX invert GPIO(78) |
| SERIAL5 | USART1 | SBUS input | |
| SERIAL6 | UART7 | Debug UART | |

Note: Some vendor silkscreens use compact labels such as
"SBUS/RSSI" around the receiver area. In ArduPilot mapping these
are separate signals: SBUS serial input is on USART1 RX (PB7),
while RSSI input is the separate digital GPIO on PC1.

## CAN

The SaamPixV1_1 exposes one CAN interface:

- CAN1 RX: PD0
- CAN1 TX: PD1

## PWM Output

The SaamPixV1_1 has 16 PWM outputs. All outputs support PWM and
DShot protocols.

| Outputs | Timer | Notes |
| --- | --- | --- |
| 1-4 | TIM1 | GPIO(50-53) |
| 5-8 | TIM4 | GPIO(54-57) |
| 9-12 | TIM3 | GPIO(58-61) |
| 13-16 | TIM2 | GPIO(62-65) |

Note: All outputs within the same timer group must use the same
protocol.

## RC Input

RC input is configured on RC/PPM input by default and supports
all ArduPilot compatible unidirectional protocols.

SBUS input is available on SERIAL5 (USART1 RX on PB7).

## GPIOs

> | GPIO | Pin | Function |

Reviewer: get rid of cpu pin column, no value to user and will confuse them

Author: Done — the CPU-pin column has been removed from the GPIO table.

| --- | --- | --- |
| GPIO(0) | PA8 | RUN_LED |
| GPIO(32) | PA7 | ALARM (TIM14) |
| GPIO(50) | PE9 | Servo1 |
| GPIO(51) | PE11 | Servo2 |
| GPIO(52) | PE13 | Servo3 |
| GPIO(53) | PE14 | Servo4 |
| GPIO(54) | PD12 | Servo5 |
| GPIO(55) | PD13 | Servo6 |
| GPIO(56) | PD14 | Servo7 |
| GPIO(57) | PD15 | Servo8 |
| GPIO(58) | PB4 | Servo9 |
| GPIO(59) | PC7 | Servo10 |
| GPIO(60) | PB0 | Servo11 |
| GPIO(61) | PB1 | Servo12 |
| GPIO(62) | PA15 | Servo13 |
| GPIO(63) | PB3 | Servo14 |
| GPIO(64) | PA2 | Servo15 |
| GPIO(65) | PA3 | Servo16 |
| GPIO(66) | PC1 | RSSI digital input |
| GPIO(78) | PB12 | UART8 TX invert (not SBUS) |

## Battery Monitor

The board has an internal voltage and current sensor. The default
battery parameters are:

> - `BATT_VOLT_PIN` = 12 (PC2)

Reviewer: get rid of cpu pins in  ()

Author: Done — the CPU pins in parentheses have been removed from the Battery Monitor section.

- `BATT_CURR_PIN` = 10 (PC0)
- `BATT_VOLT_MULT` = 10.1
- `BATT_AMP_PERVLT` = 17.0

These are set in the hwdef. Adjust `BATT_AMP_PERVLT` to match
your current sensor if using an external one.

## Compass

The SaamPixV1_1 has two built-in compasses:

> - LIS3MDL on SPI4 (ROTATION_NONE)

Reviewer: remove

Author: Done — removed.

> - AK8963 via MPU9250 on SPI4 (ROTATION_ROLL_180_YAW_90)

Reviewer: remove

Author: Done — removed.

| SERIAL2 | USART3 | Companion computer (USB) | Onboard CP2104 USB bridge |
| SERIAL3 | UART4 | GPS 1 | |
| SERIAL4 | UART8 | GPS 2 (default) | TX invert GPIO(78) |
| SERIAL5 | USART1 | User | |
| SERIAL6 | UART7 | Debug UART | |

Note: SERIAL2 (USART3) is wired to an onboard CP2104 USB-to-UART
bridge exposed on the secondary USB connector for companion-computer
connection, not to an external telemetry header. SERIAL0 (USB OTG)
is the primary USB console.

## CAN

The SaamPixV1_1 exposes one CAN interface:

- CAN1 RX: PD0
- CAN1 TX: PD1

## I2C

The external I2C connector is on the I2C2 bus (SCL on PB10, SDA on
PB11), shared with the GPS connector's I2C pins.

## PWM Output

The SaamPixV1_1 has 16 PWM outputs. All outputs support PWM and
DShot protocols.

| Outputs | Timer | Notes |
| --- | --- | --- |
| 1-4 | TIM1 | GPIO(50-53) |
| 5-8 | TIM4 | GPIO(54-57) |
| 9-12 | TIM3 | GPIO(58-61) |
| 13-16 | TIM2 | GPIO(62-65) |

Note: All outputs within the same timer group must use the same
protocol.

## RC Input

RC input is configured on the RC/PPM input by default and supports
all ArduPilot compatible unidirectional protocols, including SBUS,

> which is connected to the RC/PPM input.

Reviewer: ```suggestion
which is connected to the RC/PPM input. Bi-directional protocols, such as CRSFT/ELRS/FPort/SRLX2, need to utilize a full UART. See [RC systems](https://ardupilot.org/copter/docs/common-rc-systems.html) for more information.
```

## RSSI

RSSI is available as a digital input on GPIO(66). To use it, set
`RSSI_TYPE` = 4 and `RSSI_ANA_PIN` = 66.

## GPIOs

| GPIO | Function |
| --- | --- |
| GPIO(0) | RUN_LED |
| GPIO(32) | ALARM (TIM14) |
| GPIO(50) | Servo1 |
| GPIO(51) | Servo2 |
| GPIO(52) | Servo3 |
| GPIO(53) | Servo4 |
| GPIO(54) | Servo5 |
| GPIO(55) | Servo6 |
| GPIO(56) | Servo7 |
| GPIO(57) | Servo8 |
| GPIO(58) | Servo9 |
| GPIO(59) | Servo10 |
| GPIO(60) | Servo11 |
| GPIO(61) | Servo12 |
| GPIO(62) | Servo13 |
| GPIO(63) | Servo14 |
| GPIO(64) | Servo15 |
| GPIO(65) | Servo16 |
| GPIO(66) | RSSI digital input |

> | GPIO(78) | UART8 TX invert |

Reviewer: my previous question was about why this inverter is useful....UART inversion on an F4 is usually only for SBUS inputs when an IOMCU isnt available....I think it will be confusing to a user as to why he needs this... I would just remove it from the hwdef and readme

## Battery Monitor

The board has an internal voltage and current sensor. The default
battery parameters are:

- `BATT_VOLT_PIN` = 12
- `BATT_CURR_PIN` = 10
- `BATT_VOLT_MULT` = 10.1
- `BATT_AMP_PERVLT` = 17.0

These are set in the hwdef. Adjust `BATT_AMP_PERVLT` to match
your current sensor if using an external one.

## Compass

The SaamPixV1_1 has two built-in compasses: LIS3MDL and AK8963
(via MPU9250).

Due to potential interference from the power supply and PWM

> circuits, an external I2C compass as part of a GPS/Compass

Reviewer: ```suggestion
circuits, disabling this and using an external I2C compass as part of a GPS/Compass
```
